sklearn/utils/_param_validation.py

```diff
@@ -280,29 +281,32 @@ def __str__(self):
         return "pandas.NA"
 
 
-class StrOptions(_Constraint):
-    """Constraint representing a set of strings.
+class Options(_Constraint):
+    """Constraint representing a finite set of instances of a given type.
 
     Parameters
     ----------
-    options : set of str
-        The set of valid strings.
+    type : type
 
-    deprecated : set of str or None, default=None
-        A subset of the `options` to mark as deprecated in the repr of the constraint.
+    options : set
+        The set of valid scalars.
+
+    deprecated : set or None, default=None
+        A subset of the `options` to mark as deprecated in the string
+        representation of the constraint.
     """
 
-    @validate_params({"options": [set], "deprecated": [set, None]})
-    def __init__(self, options, deprecated=None):
+    def __init__(self, type, options, deprecated=None):
         super().__init__()
+        self.type = type
         self.options = options
         self.deprecated = deprecated or set()
 
         if self.deprecated - self.options:
             raise ValueError("The deprecated options must be a subset of the options.")
 
     def is_satisfied_by(self, val):
-        return isinstance(val, str) and val in self.options
+        return isinstance(val, self.type) and val in self.options
```

Comment:
Is this valid?

```suggestion
        return (
            isinstance(val, self.type) and
            (val in self.options or val in self.deprecated)
        ) 
```

Reply:
deprecated must be a subset of options so it would be redondant